### PR TITLE
feat(API,CLI) add support for searches with complex filters

### LINKED ISSUE #278: Logical operators in layers filters
Currently, the layer filters have an implicit & and do not allow the use of other logical operators.

### PR DESCRIPTION
**Pull request by Barbozafab**
## Description

- Created a new option for `ls` and `get` to perform searches with complex filters;

- Created new API endpoint `/api/objects/search`;

- Updated the CLI autocompleter with `cp`, `connect3d` and `disconnect3d` commands;

- Updated the layers:
  - Layers used to have an attribute `filters`, which was a `map[string]string` filled with multiple simple filters; when the layer was applied, an `AND` operation was done between all the simple filters.
  - Now, this attribute has been changed to `filter`, a single string containing a complex filter; 
  - The complex filter can be reassigned by doing `[slug]:filter=[new complex filter]`;
  - A new filter can be added to the current one, resulting in an `AND` operation between the old filter and the new one, by doing `[slug]:filter+=[new complex filter]`.

- Fixes #278 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Local test

### CODE REVIEW

File: wiki/CLI-language.md [hunk 1]
```diff
@@ -428,6 +428,14 @@ By adding the filter tag=[tag_slug] to the ls, we can obtain the children that h
 ls [path] tag=[tag_slug]
 ```
 
+#### Complex filters
+
+Complex filters can be added to the ls command with the flag `-f`, composing complex boolean expressions with the operators `=`, `!=`, `<`, `<=`, `>`, `>=`, `&` and `|`.
+
+```
```
Comment (helderbetiol) [inline wiki/CLI-language.md:435]: adding more examples would be nice, as well as mentioning the () in the text. And what about the get?

Comment (Barbozafab) [reply wiki/CLI-language.md:484]: I've rewritten this section of the Wiki to better explain the functioning of the complex filter, and added a section about it to the `get` command.

File: CLI/controllers/layer.go [hunk 2]
```diff
@@ -167,11 +168,16 @@ func (controller Controller) UpdateLayer(path string, attributeName string, valu
 	case models.LayerFiltersRemove:
 		_, err = controller.UpdateObj(path, map[string]any{attributeName: value})
 	case models.LayerFiltersAdd:
-		values := strings.Split(value.(string), "=")
-		if len(values) != 2 || len(values[0]) == 0 || len(values[0]) == 0 {
-			return fmt.Errorf("invalid filter format")
+		var filters map[string]any
+		reg := regexp.MustCompile(`^(\w(\w|\-)*)=[^!<>=&|]+$`)
+		if strings.HasPrefix(value.(string), "filter=") {
```
Comment (helderbetiol) [inline CLI/controllers/layer.go:173]: Isn't it filters+= the prefix to add filters?

File: CLI/utils/util.go [hunk 3]
```diff
@@ -255,3 +257,101 @@ func ObjectAttr(obj map[string]any, attr string) (any, bool) {
...
+
+	ops := map[string]string{"<=": "$lte", ">=": "$gte", "<": "$lt", ">": "$gt", "!=": "$ne", "=": "$eq"}
+
+	if len(expressions) <= 3 {
+		expression := strings.Join(expressions[:], "")
+
+		if match := re.FindStringSubmatch(expression); match != nil {
+			switch match[1] {
+			case "startDate":
+				// if match[2] != "=" {
```
Comment (helderbetiol) [inline CLI/utils/util.go:323]: let's erase the comments

File: CLI/controllers/commandController.go [hunk 4]
```diff
@@ -136,11 +136,17 @@ func (controller Controller) ObjectUrlGeneric(pathStr string, depth int, filters
 		params.Add("limit", strconv.Itoa(depth))
 	}
 
+	endpoint := "/api/objects"
 	for key, value := range filters {
-		params.Set(key, value)
+		fmt.Println("k, v", key, value)
```
Comment (helderbetiol) [inline CLI/controllers/commandController.go:141]: bye bye print?

**Comment by helderbetiol** [comment]
If someone forgets to put the path in the ls command, maybe we should give a better error:
```
/Physical> ls -f reservedColor=AAAAAA
Error: on GET /api/hierarchy-objects/reservedColor=AAAAAA/all?limit=1 : cannot unmarshal json : 
404 page not found
```

**Comment by helderbetiol** [comment]
The classic (simple) filter could use wildcards (*) as bellow while the complex one cannot:
 ```
 admin@TenantName:/Physical> ls . name=site*
k, v name site*
siteA
siteB
siteD
admin@TenantName:/Physical> ls . -f name=site*
k, v filter name=site*
Error: Invalid filter expression
 ```

File: CLI/controllers/layer.go (hunk 2 above)
Comment (Barbozafab) [reply CLI/controllers/layer.go:173]: It is! Line 170 (`case models.LayerFiltersAdd:`) checks for this. Line 173 checks if the user has specified `filter=` in the complex filter. 
e.g. `test:filters+=filter=category!=room`

File: API/controllers/entityController.go [hunk 5]
```diff
@@ -506,7 +507,141 @@ func HandleGenericObjects(w http.ResponseWriter, r *http.Request) {
 
 			return imageIDToUrl(u.EntityStrToInt(entityStr), object)
 		})
+		u.Respond(w, u.RespDataWrapper("successfully processed request", matchingObjects))
+	}
+}
+
+// swagger:operation POST /api/objects/search Objects GetGenericObject
+// Get all objects from any entity. Return as a list.
```
[the remark this answers is not in the document]
Comment (Barbozafab) [reply API/controllers/entityController.go]: I've updated the endpoint name on the doc and inserted info about the body of the request.

File: CLI/controllers/commandController.go (hunk 4 above)
Comment (Barbozafab) [reply CLI/controllers/commandController.go]: Bye bye print :)

**Comment by Barbozafab** [comment]
I've updated the parser so now it demands a path to be specified when performing an `ls` or `get with a complex filter, and I've also added support for wildcards on complex filters.

File: API/controllers/entityController.go [hunk 6]
```diff
@@ -506,7 +508,130 @@ func HandleGenericObjects(w http.ResponseWriter, r *http.Request) {
...
 			return imageIDToUrl(u.EntityStrToInt(entityStr), object)
 		})
+		u.Respond(w, u.RespDataWrapper("successfully processed request", matchingObjects))
+	}
+}
+
+// swagger:operation POST /api/objects/search Objects HandleComplexFilters
+// Get all objects from any entity that match the complex filter. Return as a list.
+// Wildcards can be used on any of the parameters present in query with equality and inequality operations.
+//
```
Comment (helderbetiol) [inline API/controllers/entityController.go:518]: maybe just remove the explication of wildcard from here and mention "check endpoint X for info on wildcards"

File: API/controllers/entityController.go [hunk 7]
```diff
@@ -506,7 +508,130 @@ func HandleGenericObjects(w http.ResponseWriter, r *http.Request) {
...
+	for _, entStr := range entities {
+		// Get objects
+		entData, err := models.GetManyObjectsComplex(entStr, req, filters, complexFilters, user.Roles)
+		if err != nil {
+			u.ErrLog("Error while looking for objects at "+entStr, "HandleComplexFilters", err.Message, r)
+			u.RespondWithError(w, err)
+			return
+		}
+
+		// Save entity to help delete and respond
```
Comment (helderbetiol) [inline API/controllers/entityController.go:623]: Is this copied code useful here?

File: API/models/model.go [hunk 8]
```diff
@@ -276,6 +280,91 @@ func GetManyObjects(entityStr string, req bson.M, filters u.RequestFilters, user
 	return data, nil
 }
 
+func GetManyObjectsComplex(entityStr string, req bson.M, filters u.RequestFilters, complexFilters map[string]any, userRoles map[string]Role) ([]map[string]interface{}, *u.Error) {
+	ctx, cancel := u.Connect()
+	var err error
+	var c *mongo.Cursor
+
+	var opts *options.FindOptions
```
Comment (helderbetiol) [inline API/models/model.go:288]: I think I've seen this code before, maybe we can create a function for it?

File: API/controllers/entityController.go (hunk 6 above)
Comment (Barbozafab) [reply API/controllers/entityController.go]: Done!

File: API/controllers/entityController.go (hunk 7 above)
Comment (Barbozafab) [reply API/controllers/entityController.go:668]: It is now that the `DELETE` method has been added to the `HandleComplexFilters` endpoint.

**Review by helderbetiol — APPROVED** [review]

**Pull request merged**